src/game.ml

```diff
@@ -54,13 +99,191 @@ module type SoldierLogic = sig
 end
 
 module Pawn : SoldierLogic = struct
+  (** Returns true if the previous move is en passant-able (i.e. moved
+      two from the previous y-position), and that the current coordinate
+      position of the pawn can move diagonally due to en passant rule *)
+
+  let check_en_passant_left
+      (curr_x, curr_y)
+      ((last_x_old, last_y_old), (last_x_new, last_y_new))
+      color
+      board =
+    let last_piece = board.(last_x_new).(last_y_new) in
+    let net_y = last_y_new - last_y_old in
+
+    if color = White then
+      if
+        (* Looking diagonally up-left *)
+        last_piece = Some (Black, Pawn)
+        && net_y = -2
+        && last_x_new - curr_x = -1
+        && board.(curr_x - 1).(curr_y + 1) = None
+      then true
+      else false
+    else if
+      (* Looking diagonally down-left *)
+      last_piece = Some (White, Pawn)
+      && net_y = 2
+      && last_x_new - curr_x = -1
+      && board.(curr_x - 1).(curr_y - 1) = None
+    then true
+    else false
+
+  let check_en_passant_right
+      (curr_x, curr_y)
+      ((last_x_old, last_y_old), (last_x_new, last_y_new))
+      color
+      board =
+    let last_piece = board.(last_x_new).(last_y_new) in
+    let net_y = last_y_new - last_y_old in
+
+    if color = White then
+      if
+        (* Looking diagonally up-right *)
+        last_piece = Some (Black, Pawn)
+        && net_y = -2
+        && last_x_new - curr_x = 1
+        && board.(curr_x - 1).(curr_y + 1) = None
+      then true
+      else false
+    else if
+      (* Looking diagonally down-right *)
+      last_piece = Some (White, Pawn)
+      && net_y = 2
+      && last_x_new - curr_x = 1
+      && board.(curr_x - 1).(curr_y - 1) = None
+    then true
+    else false
+
+  let is_valid_square_pawn
+      (curr_x, curr_y)
+      board
+      color
+      last_move
+      (pot_x, pot_y) : bool =
+    let net_x = pot_x - curr_x in
+    let net_y = pot_y - curr_y in
+    let left_en_passant_able =
+      check_en_passant_left (curr_x, curr_y) last_move color board
+    in
+    let right_en_passant_able =
+      check_en_passant_right (curr_x, curr_y) last_move color board
+    in
+    let basic_valid_square =
+      is_valid_square board color (pot_x, pot_y)
+    in
+
+    let check_conditions =
```

Comment:
This function has a lot of repeating code and graders are not gonna like it. I am pretty sure you can make it agnostic of our color. Using just the curr pos and the potential pos, we can determine the legality of a pawn move. Here is an example for moving two spaces forward (pseudocode):

If (net y = 2 or -2) && (current pos = 1 or 6) && (empty squares in +y or empty squares in -y) then true.
Notice, I don't even need to check net x because I know the only time net y = 2 or -2 is when net x = 0.

We can discuss this more if you want on discord.